src/ncp-spinel/SpinelNCPTaskJoinerCommissioning.cpp

```diff
@@ -0,0 +1,159 @@
+/*
+ *
+ * Copyright (c) 2018 Nest Labs, Inc.
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "assert-macros.h"
+#include <syslog.h>
+#include <errno.h>
+#include "SpinelNCPTaskJoinerCommissioning.h"
+#include "SpinelNCPInstance.h"
+#include "any-to.h"
+#include "spinel-extra.h"
+
+using namespace nl;
+using namespace nl::wpantund;
+
+nl::wpantund::SpinelNCPTaskJoinerCommissioning::SpinelNCPTaskJoinerCommissioning(
+	SpinelNCPInstance* instance,
+	CallbackWithStatusArg1 cb,
+	bool action,
+	const char *psk,
+	const char *provisioning_url
+	):	SpinelNCPTask(instance, cb), mAction(action),
+	mPsk(psk), mProvisioningUrl(provisioning_url),
+	mLastState(instance->get_ncp_state())
+{
+}
+
+void
+nl::wpantund::SpinelNCPTaskJoinerCommissioning::finish(int status, const boost::any& value)
+{
+	if (mAction && status != kWPANTUNDStatus_Ok)
+	{
+		mInstance->change_ncp_state(mLastState);
+	}
+
+	SpinelNCPTask::finish(status, value);
+}
+
+int
+nl::wpantund::SpinelNCPTaskJoinerCommissioning::vprocess_event(int event, va_list args)
+{
+	int ret = kWPANTUNDStatus_Failure;
+	Data frame;
+
+	EH_BEGIN();
+
+	if (!mInstance->mEnabled) {
+		ret = kWPANTUNDStatus_InvalidWhenDisabled;
+		finish(ret);
+		EH_EXIT();
+	}
+
+	if (mInstance->get_ncp_state() == UPGRADING) {
+		ret = kWPANTUNDStatus_InvalidForCurrentState;
+		finish(ret);
+		EH_EXIT();
+	}
+
+	// Wait for a bit to see if the NCP will enter the right state.
+	EH_REQUIRE_WITHIN(
+		NCP_DEFAULT_COMMAND_RESPONSE_TIMEOUT,
+		!ncp_state_is_initializing(mInstance->get_ncp_state()),
+		on_error
+	);
+
+	if (ncp_state_is_associated(mInstance->get_ncp_state())) {
+		ret = kWPANTUNDStatus_Already;
+		finish(ret);
+		EH_EXIT();
+	}
+
+	// The first event to a task is EVENT_STARTING_TASK. The following
+	// line makes sure that we don't start processing this task
+	// until it is properly scheduled. All tasks immediately receive
+	// the initial `EVENT_STARTING_TASK` event, but further events
+	// will only be received by that task once it is that task's turn
+	// to execute.
+	EH_WAIT_UNTIL(EVENT_STARTING_TASK != event);
+
+	if (mAction)
+	{
+		mInstance->change_ncp_state(ASSOCIATING);
```

Comment:
Let's save the `mLastState` here before changing state (since it may have changed from the time we save it from the constructor)

`mLastState = mInstance->get_ncp_state();`